### DISCUSSION TITLE
Support for network params on cmdline (like old IPAPPEND)

### DISCUSSION THREAD

**Discussion by eddyg** [ltsp/ltsp#217] [Q&A]
Is there a way in LTSP19 to pass the network parameters acquired in the iPXE pre-boot environment down to the kernel to avoid having to DHCP again?

One thing I noticed that seems to add extra time to client boots is `IP-Config`. I see output like this:
```
IP-Config: eno1 hardware addresss aa:bb:cc:dd:ee:ff mtu 1500 DHCP RARP
IP-Config: no response after 2 secs - giving up
IP-Config: eno1 hardware addresss aa:bb:cc:dd:ee:ff mtu 1500 DHCP RARP
IP-Config: no response after 3 secs - giving up
IP-Config: eno1 hardware addresss aa:bb:cc:dd:ee:ff mtu 1500 DHCP RARP
IP-Config: eno1 complete (dhcp from 192.168.1.1):
```

(FWIW, this `IP-Config` output appears even if the `quiet` option is specified as a `cmdline` parameter to suppress all the kernel noise during boot.)

In my quest to get faster client boots, I'm hoping there's some "clean" way to avoid `IP-Config` that still plays nice with the rest of Ubuntu 18.04's networking?

_P.S. Thanks for all the work on the new LTSP!_

**Comment by alkisg** [comment]
This ipconfig output suggests a problem with your DHCP server. You first need to troubleshoot that one; why does it need 4 seconds to reply?

To emulate IPAPPEND 3, put this in ltsp.conf and run `ltsp ipxe`:

```shell
[server]
POST_IPXE_SED="sed 's|cmdline_ltsp |&ip=\${ip}:\${srv}:\${gateway}:\${netmask} |' -i /srv/tftp/ltsp/ltsp.ipxe"
```

**Comment by eddyg** [comment]
Looking at the logs on the switch, link is going away between the PXE pre-boot environment and when `IP-Config` tries to run. It takes several seconds before link comes back. I was hoping maybe the switchports were not configured to be `portfast`, but they are.

I'm using two different NUC models: one has an Intel i219-V NIC (uses the `e1000e` driver) and the other has a Realtek 8111HN NIC (uses the `r8169` driver) and they *both* need multiple `IP-Config` DHCP retries. (The clients are booting Ubuntu 18.04.3 with the HWE kernel.)

I can watch the LEDs on the Ethernet jack of the clients and see link disappear during the boot. By recording a video of the console and the network port, I can see this is where link goes down on the NUC with the Realtek NIC:
``` 
kernel: [    2.872001] Generic PHY r8169-300:00: attached PHY driver [Generic PHY] (mii_bus:phy_addr=r8169-300:00, irq=IGNORE)
kernel: [    2.980145] r8169 0000:03:00.0 enp3s0: Link is Down
kernel: [    6.873339] r8169 0000:03:00.0 enp3s0: Link is Up - 1Gbps/Full - flow control off
```

but interestingly enough, on the NUC with the Intel NIC, as **soon** as `initrd.img` is done transferring, link goes away and *stays down* until over 8 seconds after the kernel has started:
```
kernel: [    1.819998] e1000e 0000:00:1f.6 0000:00:1f.6 (uninitialized): registered PHC clock
kernel: [    1.891445] e1000e 0000:00:1f.6 eth0: (PCI Express:2.5GT/s:Width x1) aa:bb:cc:dd:ee:ff
kernel: [    1.892114] e1000e 0000:00:1f.6 eth0: Intel(R) PRO/1000 Network Connection
kernel: [    1.892853] e1000e 0000:00:1f.6 eth0: MAC: 12, PHY: 12, PBA No: FFFFFF-0FF
kernel: [    1.894052] e1000e 0000:00:1f.6 eno1: renamed from eth0
...
kernel: [    8.322405] e1000e: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
```

Not sure what's going on, but it's definitely a significant source of delay for both of these clients.

**Comment by alkisg** [comment]
The previous "NIC up" state of PXE/iPXE cannot be "transferred" to the kernel as it's a completely unrelated "operating system". BIOS/PXE/iPXE have their own driver, while the Linux kernel has its own drivers for all devices and reinitializes everything from scratch.

You may also try putting "break=mount" to the kernel cmdline, so that it stops before NFS is mounted. And then, try to rmmod and modprobe the NIC drivers and see if you can easily reproduce the delay there too. It might help you see if it's a problem with the kernel, with udev or with the switch.

In any case, while some LTSP user may know networking internals and chime in and help more, the LTSP code by itself isn't related to all these at all. I.e. if you decide it's related to the kernel, you will definitely get more help from the kernel devs (mailing lists etc) than from here.

**Comment by eddyg** [comment]
Your sample `POST_IPXE_SED` works great to avoid another DHCP request from the OS. Thanks! I just needed to add a `DNS_SERVER` entry in my *ltsp.conf* since that was previously obtained via DHCP.

And to close the loop on the issue I mentioned on netboot being slow with the `e1000e` driver, it seems to just be inherent to the driver. According to someone on the e1000-devel mailing list:

> the physical interface for the NIC is powered down in `e1000e_reset` via `e1000_power_down_phy` and is not powered back on until `e1000e_open` calls `e1000_power_up_phy`

So auto-negotiation (which itself takes 3-4 seconds!) can't even _start_ until systemd brings the interface up and configures the NIC during boot. (It's too bad there's not some way for iPXE to leave at least the physical aspect of the NIC configured since auto-negotiation has already happened there!)

**Comment by alkisg** [comment]
Debian initramfs doesn't use systemd; it's udev that populates the device tree.
If you're prepared to write code to make your system faster, you could probably initialize your network card a bit sooner than needed, e.g. in init-top, so that it's ready a couple of secs earlier.

**Comment by eddyg** [comment]
You're right, I meant udev... and perhaps I already tried something similar to what you're suggesting?

I created `/etc/udev/rules.d/20-e1000e.rules` with a rule that gets executed when the `e1000e` module gets loaded:

    ACTION=="add", SUBSYSTEM=="module", KERNEL=="e1000e", RUN+="/etc/ltsp/link-up"

where `/etc/ltsp/link-up` was:

```
#!/bin/bash
until [ -d /sys/class/net/eno1 ]; do sleep 0.1; done
ip link set eno1 up >/tmp/eno1-up.out 2>&1
```

and it didn't make any noticeable difference. (In my testing with `break=mount`, simply setting the link `up` like that _was_ enough to start auto-negotiation and make the LEDs on the NIC come on).

I also found [this commit in the iPXE Intel driver](https://git.ipxe.org/ipxe.git/commit/546dd51de?f=src/drivers/net/intel.c) very interesting... (iPXE seems to deal with this NIC just fine).

**Comment by alkisg** [comment]
Try `break=top` in the cmdline, and at that point, `modprobe e1000`. Does it make a difference?
